engine/table/src/main/java/io/deephaven/engine/table/impl/MultiJoinTableImpl.java

```diff
@@ -35,19 +34,139 @@
+                                "Key column mismatch for table %d, first table has key columns=%s, this table has %s",
+                                tableNumber, inputHelper.keySourceMap.keySet(), keySourceMap.keySet()));
+            }
+            // Verify matching column types (using key order from hashmaps).
+            final Collection<Class<?>> currentColumnTypes =
+                    keySourceMap.values().stream().map(ColumnSource::getType).collect(Collectors.toSet());
+            final Collection<Class<?>> expectedColumnTypes =
+                    inputHelper.keySourceMap.values().stream().map(ColumnSource::getType).collect(Collectors.toSet());
+            if (!currentColumnTypes.equals(expectedColumnTypes)) {
+                throw new IllegalArgumentException(
+                        String.format(
+                                "Key column type mismatch for table %d, first table has key column types=%s, this table has %s",
+                                tableNumber, expectedColumnTypes, currentColumnTypes));
+            }
+            // Verify matching column component types (using key order from hashmaps).
+            final Collection<Class<?>> currentComponentTypes =
+                    keySourceMap.values().stream().map(ColumnSource::getComponentType).collect(Collectors.toSet());
+            final Collection<Class<?>> expectedComponentTypes =
+                    inputHelper.keySourceMap.values().stream().map(ColumnSource::getComponentType)
+                            .collect(Collectors.toSet());
+            if (!currentComponentTypes.equals(expectedComponentTypes)) {
+                throw new IllegalArgumentException(String.format(
+                        "Key column component type mismatch for table %d, first table has key column component types=%s, this table has %s",
+                        tableNumber, expectedComponentTypes, currentComponentTypes));
+            }
+        }
+
+        ColumnSource<?>[] keySources(final String[] columns) {
+            return Arrays.stream(columns).map(keySourceMap::get).toArray(ColumnSource<?>[]::new);
+        }
+
+        ColumnSource<?>[] keySources() {
+            return keySources(keyColumnNames);
+        }
+
+        ColumnSource<?>[] originalKeySources(final String[] columns) {
+            return Arrays.stream(columns).map(originalKeySourceMap::get).toArray(ColumnSource<?>[]::new);
+        }
+
+        ColumnSource<?>[] originalKeySources() {
+            return originalKeySources(keyColumnNames);
+        }
+    }
+
+    @VisibleForTesting
     static MultiJoinTableImpl of(@NotNull final JoinControl joinControl,
             @NotNull final MultiJoinInput... multiJoinInputs) {
-        return QueryPerformanceRecorder.withNugget("multiJoin",
-                () -> new MultiJoinTableImpl(joinControl, multiJoinInputs));
+
+        final Table[] tables = Arrays.stream(multiJoinInputs).map(MultiJoinInput::inputTable).toArray(Table[]::new);
```

Comment:
We could probably memoize this, using the memoization cache on the first table.

Reply:
This can hold for another PR.

Reply:
Created issue https://github.com/deephaven/deephaven-core/issues/4273 to track this.